packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeDefinition.cs

```diff
@@ -1311,89 +1320,121 @@ private static MethodBodyStatement ThrowValidationFailException(ValueExpression
+                if (property.WireInfo == null)
                 {
                     continue;
                 }
-                var propertySerializationName = wireInfo.SerializedName;
-                var propertySerializationFormat = wireInfo.SerializationFormat;
-                var propertyIsReadOnly = wireInfo.IsReadOnly;
-                var propertyIsRequired = wireInfo.IsRequired;
-                var propertyIsNullable = wireInfo.IsNullable;
-
-                // Generate the serialization statements for the property
-                var serializationStatement = CreateSerializationStatement(property.Type, property, propertySerializationFormat);
-
-                // Check for custom serialization hooks
-                foreach (var attribute in _model.CustomCodeView?.GetAttributes()
-                             .Where(a => a.AttributeClass?.Name == CodeGenAttributes.CodeGenSerializationAttributeName) ?? [])
+
+                propertyStatements.Add(CreateWritePropertyStatement(property));
+            }
+
+            foreach (var field in _model.CanonicalView.Fields)
+            {
+                // we should only write those properties with a wire info. Those properties without wireinfo indicate they are not spec properties.
+                if (field.WireInfo == null)
                 {
-                    if (CodeGenAttributes.TryGetCodeGenSerializationAttributeValue(
-                            attribute,
-                            out var name,
-                            out _,
-                            out var serializationHook,
-                            out _,
-                            out _) && name == property.Name && serializationHook != null)
-                    {
-                        serializationStatement = This.Invoke(
-                                serializationHook,
-                                _utf8JsonWriterSnippet,
-                                _serializationOptionsParameter)
-                            .Terminate();
-                    }
+                    continue;
                 }
 
-                var writePropertySerializationStatements = new MethodBodyStatement[]
-                {
-                    _utf8JsonWriterSnippet.WritePropertyName(propertySerializationName),
-                    serializationStatement
-                };
+                propertyStatements.Add(CreateWritePropertyStatement(field: field));
+            }
+
+            return [.. propertyStatements];
+        }
 
-                // Wrap the serialization statement in a check for whether the property is defined
-                var wrapInIsDefinedStatement = WrapInIsDefined(property, property, propertyIsRequired, propertyIsReadOnly, propertyIsNullable, writePropertySerializationStatements);
-                if (propertyIsReadOnly && wrapInIsDefinedStatement is not IfStatement)
+        private MethodBodyStatement CreateWritePropertyStatement(
```

Comment:
open question: do you think it's worth splitting this up into two separate methods, one for serializing fields and another for properties for improved readability. ie. `CreateWritePropertyStatement(property)` & `CreateWriteFieldStatement(field)` ? I guess the downside to this is repeated code.

Reply:
I was rationalizing leaving this as one method in my head because we are writing JSON "properties" which may come from a CSharp property or a field.

Reply:
Would it be better if we have a centralized method that takes those "data part" as input, and a method calling it taking a property or a field?
looks like this:
```csharp
private MethodBodyStatement CreateWritePropertyStatement(PropertyProvider? property, FieldProvider? field)
	=> CreateWritePropertyStatement(
		property?.WireInfo ?? field?.WireInfo ?? throw new Exception(),
		// other data parameters
	);

private MethodBodyStatement CreateWritePropertyStatement(WireInfo wireInfo, // other parameters) {}
```
this would make this look nicer and better handle the nullability.